File: NeetCode/18-Math-Geometry/DetectSquares.java
import java.util.*;

public class DetectSquares {
    List<int[]> pointList;
    Map<String, Integer> pointMap;

    public DetectSquares() {
        pointList = new ArrayList<>();
        pointMap = new HashMap<>();
    }
    
    public void add(int[] point) {
        pointList.add(point);
        String key = point[0] + "," + point[1];
        if(pointMap.containsKey(key)){
            pointMap.put(key, pointMap.get(key) + 1);
        }
        else{
            pointMap.put(key, 1);
        }
        
    }
    
    public int count(int[] point) {
        int totalSquare = 0;
        int x = point[0];
        int y = point[1];
        for (int[] p : pointList){
            int x2 = p[0];
            int y2 = p[1];
            if((Math.abs(x - x2) == Math.abs(y - y2)) && (x != x2)){
                String key1 = x + "," + y2;
                String key2 = x2 + "," + y;
                if(pointMap.containsKey(key1) && pointMap.containsKey(key2)){
                    int count1 = pointMap.get(key1);
                    int count2 = pointMap.get(key2);
                    totalSquare += count1 * count2;
                } 
            }
              
        }
        return totalSquare;
    }
}

File: NeetCode/18-Math-Geometry/DetectSquares.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <cmath>

using namespace std;

class DetectSquares {
public:
    vector<vector<int>> pointList;
    unordered_map<string, int> pointMap;

    DetectSquares() {
        
    }
    
    void add(vector<int> point) {
        pointList.push_back(point);
        string key = to_string(point[0]) + "," + to_string(point[1]);
        pointMap[key]++;
    }
    
    int count(vector<int> point) {
        int totalSquare = 0;
        int x = point[0];
        int y = point[1];
        for (vector<int>& p : pointList){
            int x2 = p[0];
            int y2 = p[1];
            if((abs(x - x2) == abs(y - y2)) && (x != x2)){
                string key1 = to_string(x) + "," + to_string(y2);
                string key2 = to_string(x2) + "," + to_string(y);
                if(pointMap.count(key1) > 0 && pointMap.count(key2) > 0){
                    int count1 = pointMap[key1];
                    int count2 = pointMap[key2];
                    totalSquare += count1 * count2;
                } 
            }
              
        }
        return totalSquare;
    }
};

